File: LeetCode/practice-cpp/LeetCode-206.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 反转链表
// null    1 -> 2 -> 3
// null <- 1    2 -> 3
// null <- 1 <- 2    3
// null <- 1 <- 2 <- 3
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) return head;

        ListNode* p0 = nullptr;
        ListNode* p1 = head;

        while (p1 != nullptr) { // 修改 p0 和 p1 之间的指向 
            ListNode* pt = p1->next;
            p1->next = p0;
            p0 = p1;
            p1 = pt;
        }
        return p0;
    }
};

File: LeetCode/practice-rust/LeetCode-206.rs
// 反转链表
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }


impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut p0 = None;
        let mut p1 = head;
        
        while let Some(mut node) = p1 { // node 获取了所有权
            p1 = node.next.take();
            node.next = p0;
            p0 = Some(node);
        }
        p0
    }
}